File: rack/drivers/VehicleProxy.java
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2010 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
*      Matthias Hentschel  <hentschel@rts.uni-hannover.de>
 *
 */
package rack.drivers;

import rack.main.*;
import rack.main.tims.*;
import rack.main.RackName;
import rack.drivers.VehicleSetValueMsg;

public class VehicleProxy extends RackDataProxy
{
    public static final byte MSG_VEHICLE_SET_VALUE =
        RackProxy.MSG_POS_OFFSET + 1;

    public VehicleProxy(int system, int instance, TimsMbx replyMbx)
    {
        super(RackName.create(system, RackName.VEHICLE, instance, 0), replyMbx, 500);
    }

    public synchronized VehicleDataMsg getData(int recordingTime)
    {
        try
        {
            TimsRawMsg raw = getRawData(recordingTime);

            if (raw != null)
            {
                VehicleDataMsg data = new VehicleDataMsg(raw);
                return data;
            }
            else
            {
                return null;
            }
        }
        catch(TimsException e)
        {
            System.out.println(e.toString());
            return null;
        }
    }

    public synchronized VehicleDataMsg getData()
    {
        return(getData(0));
    }

    public synchronized void setValue(VehicleSetValueMsg values)
    {
        currentSequenceNo++;
        try
        {
            replyMbx.send(MSG_VEHICLE_SET_VALUE, commandMbx,
                               (byte)0, currentSequenceNo, values);

            TimsRawMsg reply;
            do
            {
              reply = replyMbx.receive(replyTimeout);
            }
            while(reply.seqNr != currentSequenceNo);
        }
        catch(TimsException e)
        {
            System.out.println(RackName.nameString(replyMbx.getName()) + ": " +
                               RackName.nameString(commandMbx) +
                               ".setValue " + e);
        }
    }
}

File: rack/drivers/vehicle_proxy.cpp
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2010 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Matthias Hentschel  <hentschel@rts.uni-hannover.de>
 *
 */
 #include <drivers/vehicle_proxy.h>

 int VehicleProxy::getData(vehicle_data *recv_data, ssize_t recv_datalen,
                        rack_time_t timeStamp, uint64_t reply_timeout_ns)
{
    RackMessage msgInfo;

    int ret = RackDataProxy::getData((void *)recv_data, recv_datalen,
                                     timeStamp, reply_timeout_ns, &msgInfo);
    if (ret) {
        return ret;
    }

    recv_data = VehicleData::parse(&msgInfo);
    return 0;
}

int VehicleProxy::setValue(vehicle_set_value_data *recv_data, ssize_t recv_datalen,
                       uint64_t reply_timeout_ns)
{
    return proxySendDataCmd(MSG_VEHICLE_SET_VALUE, recv_data, recv_datalen,
                            reply_timeout_ns);
}
